### PR TITLE
[AIRFLOW-1057] fix class name that is changed at docker package version 2.0.0

### PR DESCRIPTION
**Pull request by wasperen**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [AIRFLOW-1057](https://issues.apache.org/jira/browse/AIRFLOW-1057) 

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
This fixes an incorrect reference to a class in the docker package

### Tests
- [X] My PR also changes the docker_operator test to check if the correct version of docker python is installed.

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by bolkedebruin** [comment]
Thanks for the contribution. Please follow the commit guidelines. Also, is this backward compatible?

**Comment by wasperen** [comment]
I have now improved the import to cater for those people that use docker-py prior to 2.0.0. It will now first try to import APIClient, but if that fails it imports Client as APIClient.

**Comment by GaelMagnan** [comment]
I've checked this pr it seems ok to me.

Maybe changing the require to be able to install either the docker-py (v1.xx) or docker (v2.xx) would be a good add.

**Comment by wasperen** [comment]
I am not familiar with the docker-py package. The package on pypi has not been updated since November 2016 and the documentation of that package seems to point to the docker package documentation (https://docker-py.readthedocs.io/en/latest/)

I have manually installed the docker package and did not rely on dependencies from setup.py. That's how I found out.

Can someone shed a light on the difference between these two packages?

**Comment by GaelMagnan** [comment]
They are basically the same package. Prior to version 2.x it was called docker-py, and since version 2 they switched to docker.
They actually have issues open on their project to make it more explicit because it's been causing a lot of confusion.

**Comment by wasperen** [comment]
Ok. I'll make the dependency such that it checks if either docker-py or docker is installed. Installing the docker dependency (in setup.py) will pull down the docker package.

**Comment by wasperen** [comment]
ok -- I know where to change the docker package dependency: line 125 of `setup.py`. Changing that to `docker = ['docker>=2.0.0']` will get us install the package docker.

But I am not sure how to validate if docker-py may be installed and rely on that version. I am new to python packaging.

**Comment by GaelMagnan** [comment]
I'm not sure if it's considered good practice, but you can import it in a try except and if it's present it won't throw.

**Comment by wasperen** [comment]
that is now the way I do it. the import tries to `import docker` and use the `APIClient` class. If that fails it imports docker-py and use the class `Client as APIClient`.
I have also changed the dependency on the docker module rather than docker-py

File: tests/operators/docker_operator.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 try:
     from airflow.operators.docker_operator import DockerOperator
-    from docker.client import Client
+    from docker.client import APIClient
```
Comment (GaelMagnan) [inline tests/operators/docker_operator.py]: You don't test if docker-py is installed during the tests to fallback?

**Review by GaelMagnan — COMMENTED** [review]
It looks good to me, just the test maybe have the same comportment that in the operator, to handle the case where the docker-py library is installed.

**Comment by wasperen** [comment]
updated the test so is does the same check on which docker package is installed.
I think this is now good to go?

**Comment by bolkedebruin** [comment]
please squash your commits, make sure you follow the commit guidelines for the commit message and then we are good to go.

**Comment by wasperen** [comment]
I think this is now in line with the guidelines -- anything I missed?

**Comment by ghost** [comment]
I think you are missing
> commits all reference JIRA issues in their subject lines

so something like `[AIRFLOW-1057] fixed class name that is changed at docker package version 2.0.0`

**Comment by wasperen** [comment]
I think it already does have a subject line that is exactly that. Do you mean something else, @avigil-twist ?

**Comment by ghost** [comment]
you changed the name of this pull request but not the name of the commit itself

**Comment by wasperen** [comment]
got it, @avigil-twist and fixed that now

**Comment by russellpierce** [comment]
This PR never got merged as far as I can tell, but the associated issue is marked as RESOLVED https://issues.apache.org/jira/browse/AIRFLOW-1057... is it actually RESOLVED?

**Comment by wasperen** [comment]
Well, @russellpierce, I must say I am not entirely sure either. I have been using my fork for most of the time. I'll check.

**Comment by wasperen** [comment]
Just had a quick look at the code (current master branch https://github.com/apache/incubator-airflow/blob/master/airflow/operators/docker_operator.py) but there we just import `Client` from `docker`. So nothing has changed there. I doubt whether this has been fixed then -- this still requires the `docker` module version < 2.0 to be installed...

I think I put the issue to "resolved" to mean: there is a pull request that needs validation. Reopened the issue.

File: airflow/operators/docker_operator.py [hunk 2]
```diff
@@ -14,12 +14,17 @@
...
 from airflow.utils.decorators import apply_defaults
 from airflow.utils.file import TemporaryDirectory
-from docker import Client, tls
-import ast
+from docker import tls
+try:
+    from docker import APIClient
+except ImportError:
+    # prior to version 2.0.0 of the module docker, the API client was called Client
+    from docker import Client as APIClient
```
Comment (ashb) [inline airflow/operators/docker_operator.py:27]: Given we now depend upon the more recent docker lib (and that is the right thing to do) I don't think there is any need to have this fallback behaviour -- we will always have at least docker 2.0 won't we?

Comment (wasperen) [reply airflow/operators/docker_operator.py:27]: Sure. Maybe we should have that reflect in the setup.py as well? The master version still read "docker = ['docker-py>=1.6.0']" which now should be "docker = ['docker-py>=2.0.0']"

**Comment by brokenjacobs** [comment]
Has this stalled again or will it make it to 2.0?

Also FYI this pr doesn't work against docker >= 3.0.0 due to breaking api changes in the docker client library:
http://docker-py.readthedocs.io/en/stable/change-log.html

**Comment by Fokko** [comment]
@brokenjacobs This first needs to be fixed; there are a lot of merge conflicts right now.

**Comment by gerardo** [comment]
This PR can be closed. It was fixed on https://github.com/apache/incubator-airflow/pull/3407

**Pull request closed without merging**